### PR TITLE
Convert sicmutils.matrix to cljs

### PR DESCRIPTION
**Pull request by sritchie**
This PR migrates `sicmutils.matrix` to cljc. Pretty easy! I just had to move some tests that were using `g/simplify` or the derivative operator up to where those are defined.

### CODE REVIEW

File: src/sicmutils/matrix.cljc [hunk 1]
```diff
@@ -0,0 +1,499 @@
...
+
+       Seqable
+       (seq [_] (seq v))
+
+       ILookup
+       (valAt [_ key] (get v key))
+       (valAt [_ key default] (get v key default))
+
+       IFn
+       (invoke [_ a]
```
Comment (sritchie) [inline src/sicmutils/matrix.cljc:72]: adding further arities is the main change here.

File: src/sicmutils/matrix.cljc [hunk 2]
```diff
@@ -0,0 +1,499 @@
...
+               (and (instance? Matrix b)
+                    (let [^Matrix bm b]
+                      (and (= r (.-r bm))
+                           (= c (.-c bm))
+                           (= v (.-v bm))))))
+       (toString [_] (str v))
+
+       Sequential
+
+       Counted
```
Comment (sritchie) [inline src/sicmutils/matrix.cljc:61]: I've also added implementations of these interfaces that structure had too.

File: src/sicmutils/matrix.cljc [hunk 3]
```diff
@@ -0,0 +1,499 @@
...
+                (Matrix. r c (mapv (fn [e] (mapv #(apply % a b c d e f g h i j k l m n o p q r s t rest) e)) v)))]))
+
+(defn matrix?
+  [m]
+  (instance? Matrix m))
+
+(defn ^:private square?
+  [^Matrix m]
+  (and (matrix? n)
+       (= (.-r m) (.-c m))))
```
Comment (sritchie) [inline src/sicmutils/matrix.cljc:199]: all `.r`, `.v` accessors change to the `.-r` form.

File: test/sicmutils/matrix_test.cljc [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+      (is (thrown? #?(:clj IllegalArgumentException :cljs js/Error) (g/* M d))))))
+
+(defn generate-square-matrix
+  [n]
+  ;; TODO Ratios aren't yet proper ratios in Clojurescript. Once we get that
+  ;; working, pull in our own gen/ratio and use that here.
+  (let [entry-gen #?(:clj gen/ratio :cljs gen/small-integer)]
+    (gen/fmap #(apply m/by-rows %) (gen/vector (gen/vector entry-gen n) n))))
+
+(defspec p+q=q+p
```
Comment (sritchie) [inline test/sicmutils/matrix_test.cljc:259]: defspec works out of the box in clojurescript!

File: test/sicmutils/matrix_test.cljc [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+; This software is distributed in the hope that it will be useful, but
+; WITHOUT ANY WARRANTY; without even the implied warranty of
+; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+; General Public License for more details.
+;
+; You should have received a copy of the GNU General Public License
+; along with this code; if not, see <http://www.gnu.org/licenses/>.
+;
+
+(ns sicmutils.matrix-test
```
Comment (sritchie) [inline test/sicmutils/matrix_test.cljc:20]: I may take another pass here and order the tests to match the implementation order in `matrix.cljc`...

**Comment by sritchie** [comment]
@littleredcomputer this is done, ready for review, and fairly straightforward!

File: test/sicmutils/matrix_test.cljc (hunk 4 above)
Comment (littleredcomputer) [reply test/sicmutils/matrix_test.cljc:259]: thank heaven

**Pull request merged**